### ISSUE TITLE
Feature request/assistance request in implementing feature: automatically generate materials and textures automatically

### ISSUE THREAD

**Issue by Tirpitz93** [overviewer/Minecraft-Overviewer#1738]
Feature: generate materials from client jar.
     this should allow using an arbitrary minecraft version and should take most the leg work out of making overviewer compatible with the latest version/snapshot. and possibly with mods that add blocks.
A few questions:

1. Currently I am not sure I understand what information is required for each material, as every block/group of blocks is handled seperately. 
2. what information comes from the map? what format numeric IDs or namespaced IDs (eg: `minecraft:blue_glazed_terracotta`, if the second is the case then it would make sense to stop using numeric IDs.

Examples of info extracted from the jar file currently.
[https://gist.github.com/Tirpitz93/31405195825cea561c48eb7c5f2f500a](https://gist.github.com/Tirpitz93/31405195825cea561c48eb7c5f2f500a)
As far as I can tell this contains all the information required to build the materials.

**Comment by DragonDev1906** [comment]
I have tried implementing this in [my fork](https://github.com/DragonDev1906/Minecraft-Overviewer/commit/da007e0d7fc2050575c5d213174a5e9faa440ed3?diff=unified).

Unfortunatly this was quite limited and although the blocks are rendered correctly (return value of the images) I wasn't able to get the images to show up on the map (see screenshot).

Additionally the code had the following limitations:

1. Only full blocks
2. Rotated Blocks are not supported
3. UV coordinates must be [0, 0, 16, 16] (can be implemented but isn't yet)
4. tintcolor (grass, leaves) (possible but isn't implemented)
5. Blocks containing multiple smaller blocks (e.g. stairs)
6. Some special handling will be required for transparent blocks (e.g. Glass) (currently we have 512 different glass versions and I have no idea how they are used)

- Partial blocks require positioning of the faces, which I didn't know whow to do correctly using PIL
- Rotated Blocks have the same issue
- Multiple smaller blocks would require knowing which block is on top.

I think the best way to continue is by having an actual renderer or by using OpenGL. All we would need would be a function that takes a list sub-blocks (coordinates inside block, size, texture for each face, rotation, (uv coordinates for each face))

**If somewone could help getting the current images to be shown in the map and to get some kind of 3D renderer as described above (or at least a working OpenGL context on headless systems together with a CPU fallback it would help a lot.**

Thanks for adding this issue, I almost forgot about that branch on my fork.


Note to the screenshot: 
Stone should be rendered automaticallt but for some reason it is called "smooth_stone" in [the jar](https://github.com/DragonDev1906/Minecraft-Overviewer/blob/automated-block-gen/overviewer_core/texturegen.py#L162) , but "stone" in [world.py](https://github.com/DragonDev1906/Minecraft-Overviewer/blob/automated-block-gen/overviewer_core/world.py#L300)

**Comment by Tirpitz93** [comment]
I was hoping to not redo the rendering side of things, I was hoping to just replace the `@material` decorated definitions with a generator.

@DragonDev1906  are you on the IRC channel or other comms platform?

**Comment by DragonDev1906** [comment]
@Tirpitz93 I usually use Discord (DragonDev1906#7588) but I can come to the IRC channel, too if you want.

**Comment by IncredibleHolg** [comment]
@DragonDev1906 
In #1613 I need to build the campfire textures from the JSON files. Do you have a tool capable of parsing the JSON structure and spitting out some python code which can then be added ``overviewer_core/textures.py``?
It does not need to be perfect, having a starting point which can then be optimized would be helpful,

**Comment by DragonDev1906** [comment]
@IncredibleHolg 
Yes and no. In the [modernGLRender](https://github.com/DragonDev1906/Minecraft-Overviewer/blob/modernGLRender/overviewer_core/texturegen.py) branch the JSON structure is parsed. 
You should be able to use `def render_blockstates(self, data: dict) -> (str, List[tuple]):` ([Line 538](https://github.com/DragonDev1906/Minecraft-Overviewer/blob/modernGLRender/overviewer_core/texturegen.py#L538)). 

It parses the JSON structure and calles the `def render_cube(self, face_texture_ids: List[int], face_uvs: list, *,
                    pos=(0, 0, 0), model_rot=(0, 0, 0), scale=(1, 1, 1), uvlock=False,
                    rotation_origin=(0, 0, 0), rotation_axis=(1, 0, 0), rotation_angle=0,
                    face_rotation=(0, 0, 0, 0, 0, 0), face_tintindex=(0, 0, 0, 0, 0, 0)):` [Line 326](https://github.com/DragonDev1906/Minecraft-Overviewer/blob/modernGLRender/overviewer_core/texturegen.py#L326) for each element (cube) defined.

This method contains all Information stored in the JSON file and only has to draw a cube with the settings given. Based on the current code in textures.py it is probably best to ignore some of these settings (e.g. face_tintindex, rotation_axis, rotation_angle, rotation_origin) and only use the ones we can actually draw using the existing rendering methods. 

Here are some things that would probably have to be changed:

- `def _render_model_to_img(self, settings: dict)` would need to create the output texture and pass it down to render_cube and return it at the end
- You may want to change `render_blockstates`, it currently returns a list of all possible variants of the block and in which NBT cases they should be applied
- Additionally you would need to copy the [load_and_combine_model](https://github.com/DragonDev1906/Minecraft-Overviewer/blob/modernGLRender/overviewer_core/asset_loader.py#L387) method and what it uses. This preprocesses the JSON file and combines the fields defined there. (it works on the `assets/minecraft/models` JSON files).

Note: It might be helpful to have a method in textures.py that can place a face at a specific position and rotation like specified with the parameters shown above.

Unfortunately I had no idea how to draw smaller cubes, other code in textures.py always cropped the texture and pasted as if it was a full cube. The numbers used there probably came from trying out what looked best. I wasn't able to figure out how to calculate the positions for how exactly to crop and paste the textures to get what I wanted. 

Since there is the issue of overlapping cubes (which isn't solved by this code but was solved via Rasterization) it would probably best to (as you suggested) making a tool that outputs python code. In that case printing in `render_cube` would probably best and would then only require a manual reordering of the lines. But you would still need to have a method for rendering a face of a smaller cube at a location with a rotation.

I hope it helps, such a tool would be helpful for all parial blocks. Keep in mind that it doesn't work for all JSON files (some contain no data, others don't use the variants structure). I checked: for the campfire it should work. If it doesn't work for a block it throws an error.
